src/indexer.c

```diff
@@ -126,59 +130,113 @@ int CountPredicateType(QueryPredicateList *queries, PredicateType type) {
     return count;
 }
 
-void indexUnderKey(INDEXER_OPERATION_T op, RedisModuleString *key, RedisModuleString *ts_key) {
+static inline void labelsIndexRemoveTsKey(RedisModuleDict *leaf,
+                                          RedisModuleString *key,
+                                          RedisModuleString *ts_key,
+                                          RedisModuleDict *_labelsIndex) {
+    RedisModule_DictDel(leaf, ts_key, NULL);
+    if (RedisModule_DictSize(leaf) == 0) {
+        RedisModule_FreeDict(NULL, leaf);
+        RedisModule_DictDel(_labelsIndex, key, NULL);
+    }
+}
+
+void labelIndexUnderKey(INDEXER_OPERATION_T op,
+                        RedisModuleString *key,
+                        RedisModuleString *ts_key,
+                        RedisModuleDict *_labelsIndex,
+                        RedisModuleDict *_tsLabelIndex) {
     int nokey = 0;
-    RedisModuleDict *leaf = RedisModule_DictGet(labelsIndex, key, &nokey);
+    RedisModuleDict *leaf = RedisModule_DictGet(_labelsIndex, key, &nokey);
     if (nokey) {
         leaf = RedisModule_CreateDict(NULL);
-        RedisModule_DictSet(labelsIndex, key, leaf);
+        RedisModule_DictSet(_labelsIndex, key, leaf);
+    }
+
+    RedisModuleDict *ts_leaf = RedisModule_DictGet(_tsLabelIndex, ts_key, &nokey);
+    if (nokey) {
+        ts_leaf = RedisModule_CreateDict(NULL);
+        RedisModule_DictSet(_tsLabelIndex, ts_key, ts_leaf);
     }
 
-    if (op == Indexer_Add) {
+    if (op & Indexer_Add) {
         RedisModule_DictSet(leaf, ts_key, NULL);
-    } else if (op == Indexer_Remove) {
-        RedisModule_DictDel(leaf, ts_key, NULL);
-        if (RedisModule_DictSize(leaf) == 0) {
-            RedisModule_FreeDict(NULL, leaf);
-            RedisModule_DictDel(labelsIndex, key, NULL);
-        }
+        RedisModule_DictSet(ts_leaf, key, NULL);
+    } else if (op & Indexer_Remove) {
+        labelsIndexRemoveTsKey(leaf, key, ts_key, _labelsIndex);
```

Comment:
@OfirMos seems like dead code because we now have `RemoveIndexedMetric_generic`. if so lets remove the INDEXER_OPERATION_T enum in the input.

Reply:
my bad, after reading the rest of the code i understand it better.